Deadlock under ~PluginView() with PDFPlugin.

lericaa opened the pull request:
#### 169eef8ec91a13161640e1bd812dcd7146a82d12
<pre>
Deadlock under ~PluginView() with PDFPlugin.
<a href="https://rdar.apple.com/108489643">rdar://108489643</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=268536">https://bugs.webkit.org/show_bug.cgi?id=268536</a>

Reviewed by NOBODY (OOPS!).

dataProviderGetBytesAtPosition might be invoked recursively from CG, and it highly increased the possiblity when the main runloop is destructing the PDFPlugin,
while the another main runloop is dispatched from dataProviderGetBytesAtPosition and does not get chance to signal semaphore as it is waiting current runloop to finish,
that causes deadlock. This change is to stop dispatch main runloop when plugin has been destroyed and use notification observer to signal the semaphore,
before main thread calling waitForCompletion for m_pdfThread.

* LayoutTests/compositing/plugins/pdf/pdf-plugin-hang-during-destruction-expected.txt: Added.
* LayoutTests/compositing/plugins/pdf/pdf-plugin-hang-during-destruction.html: Added.
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
(WebKit::pdfThreadIsExitingNotificationName):
(WebKit::PDFIncrementalLoader::clear):
(WebKit::PDFIncrementalLoader::hasBeenDestroyed const):
(WebKit::PDFIncrementalLoader::dataProviderGetBytesAtPosition):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::hasBeenDestroyed const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/169eef8ec91a13161640e1bd812dcd7146a82d12

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41725 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44107 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44300 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37818 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23906 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18070 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34485 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42299 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17694 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35932 "Found 3 new API test failures: TestWebKitAPI.WKWebView.PageZoomAfterPDF, TestWebKitAPI.WebKit.PDFLinkReferrer, TestWebKitAPI.PDF.PrintSize (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15156 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45688 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37938 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37304 "Found 9 new test failures: accessibility/mac/bounds-for-range-crash.html, accessibility/mac/iframe-relative-frame.html, accessibility/mac/search-predicate-container-not-included.html, fast/dom/object-load-pdf-data-url.html, fast/replaced/pdf-as-embed-with-no-mime-type-is-not-blank.html, fast/replaced/pdf-iframe-load-crash.html, http/tests/cache/disk-cache/shattered-deduplication.html, imported/w3c/web-platform-tests/html/semantics/embedded-content/the-object-element/object-param-url.html, platform/mac-wk2/plugins/script-object-access-fails-during-slow-initialization.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41027 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16540 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13579 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39459 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18159 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/36232 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18216 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17803 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
```diff
@@ -38,12 +38,15 @@
 namespace WebKit {
 using namespace WebCore;
 
+#define PFDThreadIsExitingNotification @"PFDThreadIsExiting_"
```
Typo
```suggestion
#define PDFThreadIsExitingNotification @"PDFThreadIsExiting_"
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
```diff
@@ -294,6 +297,7 @@
     // we can force the PDFThread to complete quickly
     if (m_pdfThread) {
         unconditionalCompleteOutstandingRangeRequests();
+        [NSNotificationCenter.defaultCenter postNotificationName:[PFDThreadIsExitingNotification stringByAppendingString:@(m_pdfThread->uid()).stringValue] object:nil];
```
Maybe write a helper? Since you have to do the same work in `ensureDataBufferLength `.

```cpp
String PDFIncrementalLoader::threadIsExitingNotificationName() const
{
    static constexpr ASCIILiteral threadIsExitingNotificationNameBase = "PFDThreadIsExiting_"s;
    return makeString(threadIsExitingNotificationNameBase, m_pdfThread->uid());
}
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
```diff
@@ -649,18 +664,32 @@ static void dataProviderReleaseInfoCallback(void* info)
     });
 
+    // One time notification observer to signal PDF thread to not wait for task dispatched in main runloop when main runloop is destructing the plugin.
+    NSNotificationCenter * __weak center = [NSNotificationCenter defaultCenter];
```
We don't need a weak reference here

lericaa replied:
hmm, why?

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
```diff
@@ -649,18 +664,32 @@ static void dataProviderReleaseInfoCallback(void* info)
+    id token = [center addObserverForName:[PFDThreadIsExitingNotification stringByAppendingString:@(Thread::current().uid()).stringValue] object:nil queue:nil usingBlock:[&dataSemaphore, &center, &token](NSNotification *) {
+        dataSemaphore.signal();
+        [center removeObserver:token];
+    }];
```
We can't pass a C++ lambda directly to an argument expecting a block.

aprotyas replied:
Let's rewrite this portion as

```suggestion
    RetainPtr<NSObject> token = [center addObserverForName:threadIsExitingNotificationName() object:nil queue:nil usingBlock:makeBlockPtr([&dataSemaphore, &center, &token](NSNotification *) {
        dataSemaphore.signal();
        [center removeObserver:token];
    }).get()];
```

lericaa replied:
🤔 i saw some other places implemented something similar, wondering if why is the difference? (sorry still a starter for c++)
i.e.:
```
m_sheetWindowWillCloseObserver = [[NSNotificationCenter defaultCenter] addObserverForName:NSWindowWillCloseNotification object:m_sheetWindow.get() queue:nil usingBlock:[weakThis = ThreadSafeWeakPtr { *this }] (NSNotification *) {
```

cdumez replied:
As far as I know, as long as it builds, it's fine. I think we only need a block in cases where we capture non-copyable stuff. @aprotyas What's the concern here since it builds?

cdumez replied:
The proposal makes the code more complicated so it needs a good reason.

aprotyas replied:
Yeah my bad about this suggestion, I probably didn't think it through. I had written some test code like this a while back that caused a runtime crash because I was providing a plain lambda and not a BlockPtr

aprotyas replied:
But that might have been specific to my use case. Since it builds and runs fine for you, keep it as is

lericaa replied:
Thanks! But you are right for making `token` as `RetainPtr<NSObject>`, as we need to retain it until destruction.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
```diff
@@ -321,6 +329,17 @@
+{
+    ASSERT(isMainRunLoop());
+
+    RefPtr plugin = m_plugin.get();
```
Can be written a little more simply:
```
RefPtr plugin = m_plugin.get();
return !plugin || plugin->hasBeenDestroyed();
```

lericaa committed:
Deadlock under ~PluginView() with PDFPlugin.
rdar://108489643
https://bugs.webkit.org/show_bug.cgi?id=268536

Reviewed by NOBODY (OOPS!).

dataProviderGetBytesAtPosition might be invoked recursively from CG, and it highly increased the possiblity when the main runloop is destructing the PDFPlugin,
while the another main runloop is dispatched from dataProviderGetBytesAtPosition and does not get chance to signal semaphore as it is waiting current runloop to finish,
that causes deadlock. This change is to stop dispatch main runloop when plugin has been destroyed and use notification observer to signal the semaphore,
before main thread calling waitForCompletion for m_pdfThread.

* LayoutTests/compositing/plugins/pdf/pdf-plugin-hang-during-destruction-expected.txt: Added.
* LayoutTests/compositing/plugins/pdf/pdf-plugin-hang-during-destruction.html: Added.
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
(WebKit::pdfThreadIsExitingNotificationName):
(WebKit::PDFIncrementalLoader::clear):
(WebKit::PDFIncrementalLoader::hasBeenDestroyed const):
(WebKit::PDFIncrementalLoader::dataProviderGetBytesAtPosition):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::hasBeenDestroyed const):
